src/nemos/type_casting.py

```diff
@@ -0,0 +1,331 @@
+    objects.
+
+    Parameters
+    ----------
+    args:
+        Any positional argument.
+    kwargs:
+        Any keyword argument.
+
+    Returns
+    -------
+    :
+        Indicates whether time-related information is consistent.
+    """
+    return _has_same_time_axis(*args, **kwargs) and _has_same_support(*args, **kwargs)
+
+
+def _get_time_info(*args, **kwargs):
+    """
+    Extract time axis and support information from the first pynapple object.
+
+    Assumes the presence of at least one pynapple object among the inputs and retrieves its time-related information.
+
+    Parameters
+    ----------
+    args:
+        Any positional argument.
+    kwargs:
+        Any keyword argument.
+
+    Returns
+    -------
+    :
+        Time axis and support information of the first pynapple object detected.
+    """
+    # list of bool
+    flat, _ = jax.tree_util.tree_flatten(jax.tree_map(is_pynapple_tsd, (args, kwargs)))
+
+    idx = flat.index(True)
+
+    # get the corresponding tsd
+    tsd, _ = jax.tree_util.tree_flatten((args, kwargs))
+
+    return tsd[idx].t, tsd[idx].time_support
+
+
+def cast_to_pynapple(
+    array: jnp.ndarray, time: NDArray, time_support: nap.IntervalSet
+) -> Union[nap.Tsd, nap.TsdFrame, nap.TsdTensor]:
+    """
+    Convert an array to a pynapple time series object.
+
+    Depending on the dimensionality of the array, creates an appropriate pynapple time series data structure using
+    the provided time and support information.
+
+    Parameters
+    ----------
+    array:
+        Data array to convert.
+    time:
+        Time axis for the pynapple object.
+    time_support:
+        Time support information for the pynapple object.
+
+    Returns
+    -------
+    :
+     A pynapple time series object based on the input array.
+    """
+    time = np.asarray(time)
+    array = np.asarray(array)
+    if array.ndim == 1:
+        return nap.Tsd(t=time, d=array, time_support=time_support)
+    elif array.ndim == 2:
+        return nap.TsdFrame(t=time, d=array, time_support=time_support)
+    else:
+        return nap.TsdTensor(t=time, d=array, time_support=time_support)
+
+
+def jnp_asarray_if(
+    x: Any, condition: Callable[[Any], bool] = is_numpy_array_like
+) -> Any:
+    """
+    Conditionally convert an object to a JAX array.
+
+    Applies the conversion if the specified condition is met. Allows for flexible handling of inputs that should
+    be treated as arrays for numerical computations.
+
+    Parameters
+    ----------
+    x:
+        Object to potentially convert.
+    condition:
+        A callable that determines whether conversion should occur.
+
+    Returns
+    -------
+    :
+        The original object or its conversion to a JAX array, based on the condition.
+    """
+    if condition(x):
+        x = jnp.asarray(x)
+    return x
+
+
+def cast_jax(func: Callable) -> Callable:
```

Comment:
should this be "cast_jax", what we're actually doing is adding pynapple support. so we should namecheck pynapple. maybe "cast_jax_pynapple" or "support_pynapple"?

Reply:
I went for support_pynapple, cause it is the most explicit, I decorated the basis evaluate as well and added tests